devices/src/virtio/balloon.rs

```diff
@@ -1,245 +1,363 @@
-// Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
 // Use of this source code is governed by a BSD-style license that can be
-// found in the LICENSE file.
-use std;
+// found in the THIRD-PARTY file.
+
+use super::super::Error as DeviceError;
+use super::{
+    ActivateError, ActivateResult, EpollConfigConstructor, Queue, VirtioDevice, TYPE_BALLOON,
+};
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use epoll;
+use logger::{Metric, METRICS};
+use memory_model::{GuestAddress, GuestMemory};
 use std::cmp;
-use std::fmt::{self, Display};
 use std::io::Write;
 use std::os::unix::io::{AsRawFd, RawFd};
 use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::mpsc;
 use std::sync::Arc;
-use std::thread;
-use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
-use msg_socket::MsgReceiver;
-use sys_util::{
-    self, error, info, warn, EventFd, GuestAddress, GuestMemory, PollContext, PollToken,
-};
-use vm_control::{BalloonControlCommand, BalloonControlResponseSocket};
-use super::{
-    DescriptorChain, Queue, VirtioDevice, INTERRUPT_STATUS_CONFIG_CHANGED,
-    INTERRUPT_STATUS_USED_RING, TYPE_BALLOON, VIRTIO_F_VERSION_1,
-};
-#[derive(Debug)]
-pub enum BalloonError {
-    /// Request to adjust memory size can't provide the number of pages requested.
-    NotEnoughPages,
-    /// Failure wriitng the config notification event.
-    WritingConfigEvent(sys_util::Error),
-}
-pub type Result<T> = std::result::Result<T, BalloonError>;
-impl Display for BalloonError {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        use self::BalloonError::*;
-        match self {
-            NotEnoughPages => write!(f, "not enough pages"),
-            WritingConfigEvent(e) => write!(f, "failed to write config event: {}", e),
-        }
-    }
-}
+use sys_util::EventFd;
+use {DeviceEventT, EpollHandler};
+
 // Balloon has three virt IO queues: Inflate, Deflate, and Stats.
 // Stats is currently not used.
 const QUEUE_SIZE: u16 = 128;
 const QUEUE_SIZES: &[u16] = &[QUEUE_SIZE, QUEUE_SIZE];
 const VIRTIO_BALLOON_PFN_SHIFT: u32 = 12;
+
+const INTERRUPT_STATUS_USED_RING: u32 = 0x1;
```

Comment:
These are already declared under different names in virtio/mod.rs

```rust
/// Interrupt flags (re: interrupt status & acknowledge registers).
/// See linux/virtio_mmio.h.
pub const VIRTIO_MMIO_INT_VRING: u32 = 0x01;
pub const VIRTIO_MMIO_INT_CONFIG: u32 = 0x02;
```

The nomenclature was changed for the other virtio devices to match the `const`s in the kernel, instead of the ones in crosvm.